### PR TITLE
Always shift times to UTC, and render w/ 3 digits of millis

### PR DESCRIPTION
**Pull request by kshakir**

### CODE REVIEW

**Comment by kshakir** [comment]
Starts writing dates in a [consistent format](https://github.com/broadinstitute/cromwell/issues/2743#issuecomment-481816760). Doesn't address old dates already converted-then-stored as strings in the database.

**Comment by cjllanwarne** [comment]
Can we link any issues to this (or is it general good-natured tidying-up?)

I'm pretty sure I've seen at least one or two time format bugs float past at various points

File: cromwellApiClient/src/main/scala/cromwell/api/model/TimeUtil.scala [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    * https://github.com/openjdk/jdk/blob/jdk8-b120/jdk/src/share/classes/java/time/OffsetDateTime.java#L1886
+    */
+  private val Iso8601MillisecondsFormat = DateTimeFormatter.ofPattern("uuuu-MM-dd'T'HH:mm:ss.SSSXXXXX")
+
+  implicit class EnhancedOffsetDateTime(val offsetDateTime: OffsetDateTime) extends AnyVal {
+    /**
+      * Discards the original timezone and shifts the time to UTC, then returns the ISO-8601 formatted string with
+      * exactly three digits of milliseconds.
+      */
+    def toUtcMilliString: String = Option(offsetDateTime).map(
```
Comment (cjllanwarne) [inline cromwellApiClient/src/main/scala/cromwell/api/model/TimeUtil.scala]: Why do we need two copy/pastes of this code?

File: cromwellApiClient/src/test/scala/cromwell/api/model/CromwellQueryResultJsonFormatterSpec.scala [hunk 2]
```diff
@@ -21,21 +33,20 @@ class CromwellQueryResultJsonFormatterSpec extends FlatSpec with Matchers {
                        |      "name": "switcheroo",
                        |      "id": "bee51f36-396d-4e22-8a81-33dedff66bf6",
                        |      "status": "Failed",
-                       |      "end": "2017-07-24T14:44:34.010-04:00",
-                       |      "start": "2017-07-24T14:44:33.227-04:00"
+                       |      "end": "2017-07-24T14:44:34.010Z",
+                       |      "start": "2017-07-24T14:44:33.227Z"
```
Comment (cjllanwarne) [inline cromwellApiClient/src/test/scala/cromwell/api/model/CromwellQueryResultJsonFormatterSpec.scala]: Weird but possibly correct? The time zone has changed by 4 hours but the hour field remains the same?

File: cromwellApiClient/src/test/scala/cromwell/api/model/CromwellQueryResultJsonFormatterSpec.scala [hunk 3]
```diff
@@ -11,8 +11,20 @@ class CromwellQueryResultJsonFormatterSpec extends FlatSpec with Matchers {
...
+      Failed,
+      OffsetDateTime.parse("2017-07-24T14:44:34.010Z"),
+      OffsetDateTime.parse("2017-07-24T14:44:33.227Z")
+    ),
+    CromwellQueryResult(
+      "switcheroo",
+      WorkflowId.fromString("0071495e-39eb-478e-bc98-8614b986c91e"),
+      Succeeded,
+      OffsetDateTime.parse("2017-07-24T15:06:45.940Z"),
+      OffsetDateTime.parse("2017-07-24T15:04:54.372Z")
```
Comment (cjllanwarne) [inline cromwellApiClient/src/test/scala/cromwell/api/model/CromwellQueryResultJsonFormatterSpec.scala]: Do we want to be able to to test "we can parse old formats in the database and convert them into the new format for display?"

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreSubmitActor.scala [hunk 4]
```diff
@@ -166,7 +166,7 @@ final case class WorkflowStoreSubmitActor(store: WorkflowStore, serviceRegistryA
 
     processSource(_.clearEncryptedValues)(originalSourceFiles) map { sourceFiles =>
       val submissionEvents: List[MetadataEvent] = List(
-        MetadataEvent(MetadataKey(id, None, WorkflowMetadataKeys.SubmissionTime), MetadataValue(OffsetDateTime.now.toString)),
+        MetadataEvent(MetadataKey(id, None, WorkflowMetadataKeys.SubmissionTime), MetadataValue(OffsetDateTime.now)),
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreSubmitActor.scala]: `. toUtcMilliString `?

File: common/src/main/scala/common/util/TimeUtil.scala [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+import java.time.format.DateTimeFormatter
+import java.time.{OffsetDateTime, ZoneOffset}
+
+object TimeUtil {
+  /**
+    * Instead of "one of" the valid ISO-8601 formats, standardize on this one:
+    * https://github.com/openjdk/jdk/blob/jdk8-b120/jdk/src/share/classes/java/time/OffsetDateTime.java#L1886
+    */
+  private val Iso8601MillisecondsFormat = DateTimeFormatter.ofPattern("uuuu-MM-dd'T'HH:mm:ss.SSSXXXXX")
```
Comment (cjllanwarne) [inline common/src/main/scala/common/util/TimeUtil.scala]: Is `XXXXX` always a constant `Z` in the render? That seems a weird way of specifying it but sure.

FWIW http://userguide.icu-project.org/formatparse/datetime lists 3 possible output formats for `XXXXX` so I guess I'm still utterly perplexed by format strings 🤷‍♂

File: cromwellApiClient/src/test/scala/cromwell/api/model/CromwellQueryResultJsonFormatterSpec.scala (hunk 3 above)
Comment (kshakir) [reply cromwellApiClient/src/test/scala/cromwell/api/model/CromwellQueryResultJsonFormatterSpec.scala]: AFAICT Cromwell never parses old-nor-new formats from the database. JSON doesn't have a `datetime` type, so we convert everything to strings. Then when reading out metadata we convert those strings... to strings, not dates.

File: cromwellApiClient/src/test/scala/cromwell/api/model/CromwellQueryResultJsonFormatterSpec.scala (hunk 2 above)
Comment (kshakir) [reply cromwellApiClient/src/test/scala/cromwell/api/model/CromwellQueryResultJsonFormatterSpec.scala]: I changed this by hand just and just turned `-04:00` to `Z`. I think it's just the GitHub word-highlighting that also selected the milliseconds.

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreSubmitActor.scala (hunk 4 above)
Comment (kshakir) [reply engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreSubmitActor.scala]: One should not be calling `Object.toString()` on **any** Metadata values as the types are not `value: String` but are `value: Any`.

Meanwhile in this PR, Metadata's `Any` converter has been updated to convert `OffsetDateTime` to the correct format. This should help in the future whenever `OffsetDateTime` instances are passed as metadata values.

File: cromwellApiClient/src/main/scala/cromwell/api/model/TimeUtil.scala (hunk 1 above)
Comment (kshakir) [reply cromwellApiClient/src/main/scala/cromwell/api/model/TimeUtil.scala]: `cromwell-api-client` despite being in the multi-project does not use project dependencies, including [`common`](https://github.com/broadinstitute/cromwell/commit/6eb104b44a03eed972c886e28a95cefa4632d019#diff-c01844a8d491e55a2a182f351df7f99c) where the other `TimeUtil` was added in this PR.

So like [`WorkflowStatus`](https://github.com/broadinstitute/cromwell/blob/40/cromwellApiClient/src/main/scala/cromwell/api/model/WorkflowStatus.scala#L5-L10), CAC needs its own copy of the util.

File: CHANGELOG.md [hunk 6]
```diff
@@ -17,6 +17,25 @@ disks: "/some/mnt 20 SSD"
 ```
 Because Cromwell's AWS backend auto-sizes disks, the size specification is simply discarded.
 
+### Time Formatting
+
+In previous versions of Cromwell, times were converted to strings using
+[the default Java formatter](https://docs.oracle.com/javase/8/docs/api/java/time/OffsetDateTime.html#toString--) which
+generates a variety of ISO-8601 formats. Strings conversions also retained whatever server time zone generated that
```
Comment (mcovarr) [inline CHANGELOG.md]: Strings => String

File: CHANGELOG.md [hunk 7]
```diff
@@ -17,6 +17,25 @@ disks: "/some/mnt 20 SSD"
...
+Going forward, times stored in Cromwell metadata, and later returned via the HTTP endpoint, are now converted to UTC
+then formatted with exactly three digits of milliseconds.
+
+For example:
+- `2017-01-19T12:34:56-04:00` will now be formatted as
+- `2017-01-19T16:34:56.000Z`
+
+This change only affects newly formatted dates. Older dates already formatted and stored by previous versions of
+Cromwell will not be updated however they will still return a
+[valid ISO-8601 format](https://en.wikipedia.org/wiki/ISO_8601). The older format may or may not include microseconds,
```
Comment (mcovarr) [inline CHANGELOG.md]: do you mean micro here?

File: cromwellApiClient/src/main/scala/cromwell/api/model/TimeUtil.scala (hunk 1 above)
Comment (mcovarr) [reply cromwellApiClient/src/main/scala/cromwell/api/model/TimeUtil.scala]: 🤦‍♂

File: common/src/main/scala/common/util/TimeUtil.scala (hunk 5 above)
Comment (kshakir) [reply common/src/main/scala/common/util/TimeUtil.scala]: Yes.

The `XXXXXX` is actually just a signal to use the constant `"Z"`, by entering this branch of code matching on one-to-six uppercase `'X'` characters, and then calling `appendOffset(…, "Z")`:

https://github.com/openjdk/jdk/blob/7217c26dc17110b3070975f7f52afde9a47d8c6a/jdk/src/share/classes/java/time/format/DateTimeFormatterBuilder.java#L1636-L1640

The second argument to `appendOffset()` is the "noOffsetText". Earlier the offset has already been removed via `.atZoneSameInstant(ZoneOffset.UTC)`.

As an alternate formatting example, using one-to-six lowercase `'x'` characters would have triggered this code branch, calling `appendOffset(…, zero)`:

https://github.com/openjdk/jdk/blob/7217c26dc17110b3070975f7f52afde9a47d8c6a/jdk/src/share/classes/java/time/format/DateTimeFormatterBuilder.java#L1641-L1646

TL;DR Could have used just `X` instead of `XXXXXX`, but wanted to copy the exact line from:
 https://github.com/openjdk/jdk/blob/7217c26dc17110b3070975f7f52afde9a47d8c6a/jdk/src/share/classes/java/time/OffsetDateTime.java#L1886

File: CHANGELOG.md (hunk 7 above)
Comment (kshakir) [reply CHANGELOG.md]: Updated the note with an example.
